src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Int32/Int32FrozenSet.cs

```diff
@@ -21,10 +21,16 @@ internal Int32FrozenSet(HashSet<int> source) : base(EqualityComparer<int>.Defaul
             Debug.Assert(source.Count != 0);
             Debug.Assert(ReferenceEquals(source.Comparer, EqualityComparer<int>.Default));
 
+            int count = source.Count;
+            int[] entries = ArrayPool<int>.Shared.Rent(count);
+            source.CopyTo(entries);
+
             _hashTable = FrozenHashTable.Create(
-                source.ToArray(),
-                item => item,
-                (_, _) => { });
+                count,
+                index => entries[index],
+                delegate { });
+
+            ArrayPool<int>.Shared.Return(entries);
```

Comment:
I never remember exactly what the rule is when to `try-finally` around pool usages and when not to.

Should this one be `try-finally`?

Reply:
There's no requirement to return arrays to the shared pool; it just means if it would have been stored, another attempt to fetch one might be more likely to have to allocate one. Generally compared to an exception, that's not a big deal, and try/finally isn't entirely free plus it adds some visual noise, so if exceptions aren't likely, often better to just keep it simple. Here, exceptions should be very rare, e.g. an OOM, and in the face of an oom, probably better to allow memory to be collected rather than increase pressure by trying to keep the object referenced.

With cancellation modeled via exception, I'm much more likely to use try/finally around cancelable operations.